llvm/lib/Bitcode/Reader/BitcodeReader.cpp

```diff
@@ -5482,6 +5482,8 @@ Error BitcodeReader::parseFunctionBody(Function *F) {
         if (!CmpInst::isIntPredicate(PredVal))
           return error("Invalid icmp predicate");
         I = new ICmpInst(PredVal, LHS, RHS);
+        if (Record[OpNum] & (1 << bitc::PSSI_SAME_SIGN))
+          cast<ICmpInst>(I)->setSameSign(true);
```

Comment:
You can't unconditionally read this, you need to check the record length first. See how the IsFP case for FMF is handled above.

This is the cause of the test failures.

Reply:
@nikic  I am working on it now. but I don't fully understand why we should record length.
can you please give a brief explanation of what the record does?

Reply:
The bitcode writer only appends the flags record if there are flags, so the bitcode reader needs to check the number of records to determine whether there are any flags.

Reply:
So the record size needs to be greater than the number of operands this means that the record has operands and flags?

```
if (Record.size() > OpNum+1 && (Record[OpNum] & (1 << bitc::ICMP_SAME_SIGN)))
          cast<ICmpInst>(I)->setSameSign();
```

Reply:
Basically yes, but the flags will be at Record[OpNum+1] and also you'll trip the preceding `OpNum+1 != Record.size()` check if you do it like this. Instead you should check this at the same place where FMF is set.